### PR TITLE
Fix default recipients for assign group in ticket notification

### PR DESCRIPTION
**Pull request by cconard96**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

Change default recipient for the "New group in assignees" ticket notification from "Group in charge of the hardware" to "Group in charge of the ticket". This matches up with the other notifications related to new actors.

### CODE REVIEW

**Comment by cedric-anne** [comment]
Tests are failing because you did not add a migration to fix existing data.

It is dangerous to touch existing data, but in this case, unless I am wrong, `ITEM_TECH_IN_CHARGE` and `ITEM_TECH_GROUP_IN_CHARGE` are not working, as `users_id_tech` and `groups_id_tech` fields are not existing for ITIL objects. So, for this notification, "Group in charge of the hardware" target should be replaced (or dropped if "Group in charge of the ticket" already exists).

And the, for all ITIL notifications, these targets should probably removed from proposed targets, and dropped from existing data.

File: install/migrations/update_10.0.2_to_10.0.3/notifications.php [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+ * @var Migration $migration
+ */
+
+$itil_types = ['Ticket', 'Change', 'Problem'];
+$iterator = $DB->request([
+    'SELECT' => ['id'],
+    'FROM'   => 'glpi_notifications',
+    'WHERE'  => [
+        'itemtype' => $itil_types,
+        'event'    => 'assign_group',
```
Comment (cedric-anne) [inline install/migrations/update_10.0.2_to_10.0.3/notifications.php:49]: Migration should parse all notifications related to ITIL objects, as it should also remove `Notification::ITEM_TECH_IN_CHARGE` and `Notification::ITEM_USER` targets.
```suggestion
```

File: install/migrations/update_10.0.2_to_10.0.3/notifications.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        'FROM'   => 'glpi_notificationtargets',
+        'WHERE'  => [
+            'notifications_id' => $notification['id']
+        ],
+    ]);
+    $targets = [];
+    foreach ($target_iterator as $target) {
+        $targets[$target['id']] = $target['items_id'];
+    }
+    $removed_item_group = false;
```
Comment (cedric-anne) [inline install/migrations/update_10.0.2_to_10.0.3/notifications.php:65]: `Notification::ASSIGN_GROUP` target should probably be added even if there is already another target (e.g. `Requester`), but only if `Notification::ASSIGN_GROUP` is not already part of targets.

See also next suggestions.

```suggestion
    $removed_item_group = false;
    $found_assigned_group = false;
```

File: install/migrations/update_10.0.2_to_10.0.3/notifications.php [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    foreach ($targets as $target_id => $target) {
+        if ((int) $target['items_id'] === Notification::ITEM_TECH_GROUP_IN_CHARGE) {
+            $DB->deleteOrDie('glpi_notificationtargets', [
+                'id' => $target['id'],
+            ]);
+            $removed_item_group = true;
+            unset($targets[$target_id]);
+        }
```
Comment (cedric-anne) [inline install/migrations/update_10.0.2_to_10.0.3/notifications.php:73]:
```suggestion
        }
        if ((int) $target['items_id'] === Notification::ASSIGN_GROUP) {
            $found_assigned_group = true;
        }
```

File: install/migrations/update_10.0.2_to_10.0.3/notifications.php [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    }
+    // If no targets remain, add a new one with items_id = Notification::ASSIGN_GROUP
+    if (count($targets) === 0) {
```
Comment (cedric-anne) [inline install/migrations/update_10.0.2_to_10.0.3/notifications.php:76]:
```suggestion
    if ($notification['event'] === 'assign_group' && $removed_item_group && !$found_assigned_group) {
```

File: install/migrations/update_10.0.2_to_10.0.3/notifications.php [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            'notifications_id' => $notification['id']
+        ],
+    ]);
+    $targets = [];
+    foreach ($target_iterator as $target) {
+        $targets[$target['id']] = $target['items_id'];
+    }
+    $removed_item_group = false;
+    foreach ($targets as $target_id => $target) {
+        if ((int) $target['items_id'] === Notification::ITEM_TECH_GROUP_IN_CHARGE) {
```
Comment (cedric-anne) [inline install/migrations/update_10.0.2_to_10.0.3/notifications.php:67]:
```suggestion
        if (
            (int) $target['items_id'] === Notification::ITEM_TECH_GROUP_IN_CHARGE
            || (int) $target['items_id'] === Notification::ITEM_TECH_IN_CHARGE
            || (int) $target['items_id'] === Notification::ITEM_USER
        ) {
```

File: install/migrations/update_10.0.2_to_10.0.3/notifications.php [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+            $DB->deleteOrDie('glpi_notificationtargets', [
+                'id' => $target['id'],
+            ]);
+            $removed_item_group = true;
+            unset($targets[$target_id]);
```
Comment (cedric-anne) [inline install/migrations/update_10.0.2_to_10.0.3/notifications.php:72]:
```suggestion
            if ((int) $target['items_id'] === Notification::ITEM_TECH_GROUP_IN_CHARGE) {
                $removed_item_group = true;
            };
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]

File: install/migrations/update_10.0.2_to_10.0.3/notifications.php [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+            $DB->deleteOrDie('glpi_notificationtargets', [
+                'id' => $target['id'],
+            ]);
+            if ((int) $target['items_id'] === Notification::ITEM_TECH_GROUP_IN_CHARGE) {
+                $removed_item_group = true;
+            }
+        }
+        if ((int) $target['items_id'] === Notification::ASSIGN_GROUP) {
+            $found_assigned_group = true;
+        }
```
Comment (cedric-anne) [inline install/migrations/update_10.0.2_to_10.0.3/notifications.php:81]:
```suggestion
    foreach ($targets as $target_id => $items_id) {
        if (
            $items_id === Notification::ITEM_TECH_GROUP_IN_CHARGE
            || $items_id === Notification::ITEM_TECH_IN_CHARGE
            || $items_id === Notification::ITEM_USER
        ) {
            $DB->deleteOrDie('glpi_notificationtargets', [
                'id' => $target['id'],
            ]);
            if ($items_id === Notification::ITEM_TECH_GROUP_IN_CHARGE) {
                $removed_item_group = true;
            }
        }
        if ($items_id === Notification::ASSIGN_GROUP) {
            $found_assigned_group = true;
        }
```

**Review by cedric-anne — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Pull request merged**